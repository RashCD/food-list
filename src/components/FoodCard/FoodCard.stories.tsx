import type { Meta, StoryObj } from '@storybook/react';

import FoodCard from './FoodCard';

const meta = {
	title: 'Components/FoodCard',
	component: FoodCard,
	parameters: {
		layout: 'centered',
	},
	tags: ['autodocs'],
	argTypes: {
		rating: {
			control: 'number',
			description: 'The rating of the food.',
			table: {
				type: {
					summary: 'number',
				},
			},
		},
		promotion: {
			control: 'select',
			description: 'The promotion name of the food.',
			options: ['gift', 'discount', '1+1', ''],
			table: {
				type: {
					summary: 'text',
				},
			},
		},
		isNew: {
			control: 'boolean',
			description: 'The isNew of the food.',
			table: {
				type: {
					summary: 'boolean',
				},
			},
		},
		minCookTime: {
			control: 'number',
			description: 'The mininum cook time of the food.',
			table: {
				type: {
					summary: 'number',
				},
			},
		},
		maxCookTime: {
			control: 'number',
			description: 'The maximum cook time of the food.',
			table: {
				type: {
					summary: 'number',
				},
			},
		},
		name: {
			control: 'text',
			description: 'The name of the food.',
			table: {
				type: {
					summary: 'text',
				},
			},
		},
		imageUrl: {
			control: 'text',
			description: 'The image url of the food.',
			table: {
				type: {
					summary: 'text',
				},
			},
		},
	},
} satisfies Meta<typeof FoodCard>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Normal: Story = {
	args: {
		rating: 3.9508,
		promotion: null,
		isNew: false,
		minCookTime: 80,
		maxCookTime: 100,
		name: 'Niquent Drinks',
		imageUrl: 'https://source.unsplash.com/random/400x400?Drinks',
	},
};

export const WithPromotion: Story = {
	args: {
		rating: 3.4708,
		promotion: 'gift',
		isNew: false,
		minCookTime: 80,
		maxCookTime: 100,
		name: 'Niquent Drinks',
		imageUrl: 'https://source.unsplash.com/random/400x400?Drinks',
	},
};

export const WithNew: Story = {
	args: {
		rating: 2.3308,
		promotion: 'gift',
		isNew: true,
		minCookTime: 80,
		maxCookTime: 100,
		name: 'Niquent Drinks',
		imageUrl: 'https://source.unsplash.com/random/400x400?Drinks',
	},
};
