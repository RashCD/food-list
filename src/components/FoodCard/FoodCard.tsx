import { Card, CardContent } from '../Card/Card';
import Badge from '../Badge/Badge';
import Icon from '../Icon/Icon';
import starIcon from '../../assets/icons/star.svg';
import styles from './foodcard.module.css';
import clsx from 'clsx';
import giftIcon from '../../assets/icons/gift.svg';
import discountIcon from '../../assets/icons/discount.svg';

type FoodCardProps = {
	className?: string;
	imageUrl: string;
	name: string;
	isNew: boolean;
	minCookTime: number;
	maxCookTime: number;
	rating: number;
	promotion: string | null;
};

const FoodCard = ({
	name,
	imageUrl,
	rating,
	minCookTime,
	maxCookTime,
	isNew,
	promotion,
}: FoodCardProps) => {
	const formatRating = (rating: number) => {
		return rating.toFixed(1);
	};

	const promotionDisplay = (promotion: string) => {
		switch (promotion) {
			case 'gift':
				return {
					className: styles.gift,
					icon: giftIcon,
					text: 'Gift',
				};
			case 'discount':
				return {
					className: styles.discount,
					icon: discountIcon,
					text: 'Discount',
				};
			default:
				return {
					className: styles.default,
					icon: null,
					text: promotion,
				};
		}
	};

	return (
		<Card className={styles.foodCardContainer}>
			{promotion && (
				<div
					className={clsx(
						styles.foodPromotionContainer,
						promotionDisplay(promotion).className,
					)}
				>
					{promotionDisplay(promotion).icon ? (
						<Icon
							source={promotionDisplay(promotion).icon as string}
							altText={`${promotionDisplay(promotion).text} icon`}
							className={styles.promotionIcon}
						/>
					) : (
						<span>{promotionDisplay(promotion).text}</span>
					)}
				</div>
			)}
			<img
				alt={`${name} image`}
				height="150"
				src={imageUrl}
				style={{
					aspectRatio: '1/2',
					objectFit: 'cover',
				}}
				width="300"
			/>
			<CardContent>
				<div>
					<span>{name}</span>
				</div>
				<div>
					<Badge variant="secondary">
						{rating && (
							<Icon
								source={starIcon}
								altText={`${name} rating`}
								className={styles.icon}
							/>
						)}
						{formatRating(rating)}
					</Badge>
					<Badge variant="secondary">{`${minCookTime} - ${maxCookTime} min`}</Badge>
					{isNew && (
						<Badge variant="secondary" className={styles.new}>
							New
						</Badge>
					)}
				</div>
			</CardContent>
		</Card>
	);
};

export default FoodCard;
