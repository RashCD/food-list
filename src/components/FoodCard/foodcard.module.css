.foodCardContainer {
	position: relative;
	color: darkslategray;
}

.foodPromotionContainer {
	display: inline-flex;
	position: absolute;
	top: 0;
	left: 0;
	background-color: red;
	border-bottom-right-radius: 0.5rem;
	width: 3rem;
	height: 2rem;
	justify-content: center;
	align-items: center;
}

.promotionIcon {
	filter: invert(100%) sepia(100%) saturate(0%) hue-rotate(288deg)
		brightness(102%) contrast(102%);
}

.default {
	color: white;
	background-color: mediumpurple;
}

.gift {
	background-color: deepskyblue;
}

.discount {
	background-color: salmon;
	color: white;
}

.icon {
	padding: 0 0.5rem 0 0;
	width: 0.75rem;
	height: 0.75rem;
}

.new {
	color: mediumseagreen;
}
