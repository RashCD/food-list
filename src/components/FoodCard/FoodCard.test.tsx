import { render, screen } from '@testing-library/react';
import FoodCard from './FoodCard';

describe('FoodCard component', () => {
	it('should render', () => {
		render(
			<FoodCard
				name="FoodCard"
				rating={4.3}
				minCookTime={80}
				maxCookTime={100}
				isNew={false}
				promotion="gift"
				imageUrl="https://some-image-url"
			/>,
		);

		expect(screen.getByText('FoodCard')).toBeInTheDocument();
	});

	it('should display gift icon when promotion is gift', () => {
		render(
			<FoodCard
				name="FoodCard"
				rating={4.3}
				minCookTime={80}
				maxCookTime={100}
				isNew={false}
				promotion="gift"
				imageUrl="https://some-image-url"
			/>,
		);

		expect(screen.getAllByRole('img')[0]).toBeInTheDocument();
		expect(screen.getByAltText('Gift icon')).toBeInTheDocument();
		expect(screen.getAllByRole('img')[0]).toHaveAttribute(
			'src',
			'/src/assets/icons/gift.svg',
		);
	});

	it('should display food images when imageUrl is provided', () => {
		render(
			<FoodCard
				name="FoodCard"
				rating={4.3}
				minCookTime={80}
				maxCookTime={100}
				isNew={false}
				promotion="gift"
				imageUrl="https://some-image-url"
			/>,
		);

		expect(screen.getAllByRole('img')[1]).toBeInTheDocument();
		expect(screen.getByAltText('FoodCard image')).toBeInTheDocument();
		expect(screen.getAllByRole('img')[1]).toHaveAttribute(
			'src',
			'https://some-image-url',
		);
	});

	it('should display rating when rating is provided', () => {
		render(
			<FoodCard
				name="FoodCard"
				rating={4.3}
				minCookTime={80}
				maxCookTime={100}
				isNew={false}
				promotion="gift"
				imageUrl="https://some-image-url"
			/>,
		);

		expect(screen.getAllByRole('img')[2]).toBeInTheDocument();
		expect(screen.getByAltText('FoodCard rating')).toBeInTheDocument();
		expect(screen.getByText('4.3')).toBeInTheDocument();
		expect(screen.getAllByRole('img')[2]).toHaveAttribute(
			'src',
			'/src/assets/icons/star.svg',
		);
	});

	it('should be able to show cooking time', () => {
		render(
			<FoodCard
				name="FoodCard"
				rating={4.3}
				minCookTime={80}
				maxCookTime={100}
				isNew={false}
				promotion="gift"
				imageUrl="https://some-image-url"
			/>,
		);

		expect(screen.getByText('80 - 100 min')).toBeInTheDocument();
	});

	it('should be able to show New badge', () => {
		render(
			<FoodCard
				name="FoodCard"
				rating={4.3}
				minCookTime={80}
				maxCookTime={100}
				isNew={true}
				promotion="gift"
				imageUrl="https://some-image-url"
			/>,
		);

		expect(screen.getByText('New')).toBeInTheDocument();
	});

	it('should match snapshot', () => {
		const comp = render(
			<FoodCard
				name="FoodCard"
				rating={4.3}
				minCookTime={80}
				maxCookTime={100}
				isNew={true}
				promotion="gift"
				imageUrl="https://some-image-url"
			/>,
		);

		expect(comp).toMatchSnapshot();
	});
});
