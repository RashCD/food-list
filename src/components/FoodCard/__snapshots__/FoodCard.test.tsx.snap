// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FoodCard component > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_card_2f15a9 _foodCardContainer_c0fa88"
      >
        <div
          class="_foodPromotionContainer_c0fa88 _gift_c0fa88"
        >
          <img
            alt="Gift icon"
            class="_icon_fa96f1 _promotionIcon_c0fa88"
            src="/src/assets/icons/gift.svg"
          />
        </div>
        <img
          alt="FoodCard image"
          height="150"
          src="https://some-image-url"
          style="aspect-ratio: 2 / 1; object-fit: cover;"
        />
        <div
          class="_cardContent_2f15a9"
        >
          <div>
            <span>
              FoodCard
            </span>
          </div>
          <div>
            <div
              class="_badge_58bfbf _secondary_58bfbf"
            >
              <img
                alt="FoodCard rating"
                class="_icon_fa96f1 _icon_c0fa88"
                src="/src/assets/icons/star.svg"
              />
              4.3
            </div>
            <div
              class="_badge_58bfbf _secondary_58bfbf"
            >
              80 - 100 min
            </div>
            <div
              class="_badge_58bfbf _secondary_58bfbf _new_c0fa88"
            >
              New
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_card_2f15a9 _foodCardContainer_c0fa88"
    >
      <div
        class="_foodPromotionContainer_c0fa88 _gift_c0fa88"
      >
        <img
          alt="Gift icon"
          class="_icon_fa96f1 _promotionIcon_c0fa88"
          src="/src/assets/icons/gift.svg"
        />
      </div>
      <img
        alt="FoodCard image"
        height="150"
        src="https://some-image-url"
        style="aspect-ratio: 2 / 1; object-fit: cover;"
      />
      <div
        class="_cardContent_2f15a9"
      >
        <div>
          <span>
            FoodCard
          </span>
        </div>
        <div>
          <div
            class="_badge_58bfbf _secondary_58bfbf"
          >
            <img
              alt="FoodCard rating"
              class="_icon_fa96f1 _icon_c0fa88"
              src="/src/assets/icons/star.svg"
            />
            4.3
          </div>
          <div
            class="_badge_58bfbf _secondary_58bfbf"
          >
            80 - 100 min
          </div>
          <div
            class="_badge_58bfbf _secondary_58bfbf _new_c0fa88"
          >
            New
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
